### ISSUE TITLE
 [BUG] inferenceにおいて値が全く変わらない

### ISSUE THREAD

**Issue by acesulfamK** [acesulfamK/AlphaSymbol#163]
# 概要
commit id: 6f5deff3b2b9b34e5255f58b9e8bfa55d7a4768a

において、2パターンの入力データに対して出力が完全に一致してしまうことが分かった。原因を探る。

# タスク
- [x] nn.Transformerの入力までで異常がないか確かめる → 結果: 異常なし
- [x] 他の重みデータで実験してみる
- [x] attn_maskの詳細を知るべくnn.Transformerの中身を見る
    - Tried 1h30min but nest is too complex...
- [ ] Compare cross entropy of bad inference in eval py
- [ ] Batch NormalizationをPos Encodingに入れてみる
- [ ] trainingでCross Entropyの入力と出力を見て出力が0.003など小さかったときに入力値はどうか見る。

**Comment by acesulfamK** [comment]
# nn.Transformer直前の入力

transformer.pyのself.transformerの2つの引数src, tgtの値を2つの入力パターンで比較してみた。

## 入力パターン1 (eval.ipynb)
```
inputs = [(0,), (0,)]
outputs = [0, 0]
```

iter 1(単語目)
```
src[0:5, 0, 0:5] (shape is S, N, E)
tensor([[  63.1745,  -22.9044,    3.1781,   -2.0811,   18.7066],
        [  41.3693,    1.3766,   -9.0516,  -19.9439,  -15.9058],
        [ -21.2904,  -67.7368,  -52.9088,   -5.1035,   11.0077],
        [ -27.7375,  -41.6544,   -4.2544,   60.3695, -103.9784],
        [  39.7710,    0.1827,  -10.5306,  -21.2674,  -17.2563]],
       device='cuda:0')

tgt[:, 0, 0:5] (shape is T, N, E)
tensor([[ 62.8923,  39.2739, -55.8710,   0.5914,   4.5007]], device='cuda:0')
```

iter 2 (2単語目)
```
tensor([[  63.1745,  -22.9044,    3.1781,   -2.0811,   18.7066],
        [  41.3693,    1.3766,   -9.0516,  -19.9439,  -15.9058],
        [ -21.2904,  -67.7368,  -52.9088,   -5.1035,   11.0077],
        [ -27.7375,  -41.6544,   -4.2544,   60.3695, -103.9784],
        [  39.7710,    0.1827,  -10.5306,  -21.2674,  -17.2563]],
       device='cuda:0')
       
tensor([[ 62.8923,  39.2739, -55.8710,   0.5914,   4.5007],
        [-83.9665,  31.7010, -75.6492,  33.1460,   1.6751]], device='cuda:0')
```

## 入力パターン2
```
inputs = [(0,), (0,)]
outputs = [1, 1]
```

iter 1
```
src[0:5, 0, 0:5] (shape is S, N, E)
tensor([[  63.1745,  -22.9044,    3.1781,   -2.0811,   18.7066],
        [  41.3693,    1.3766,   -9.0516,  -19.9439,  -15.9058],
        [ -21.2904,  -67.7368,  -52.9088,   -5.1035,   11.0077],
        [ -27.7375,  -41.6544,   -4.2544,   60.3695, -103.9784],
        [  40.6276,   11.5976,    5.9365,   66.9380,   47.0617]],
       device='cuda:0')

tgt[:, 0, 0:5] (shape is T, N, E)
tensor([[ 62.8923,  39.2739, -55.8710,   0.5914,   4.5007]], device='cuda:0')
```

iter 2
```
src same as iter 1 src

tgt
tensor([[ 62.8923,  39.2739, -55.8710,   0.5914,   4.5007],
        [-83.9665,  31.7010, -75.6492,  33.1460,   1.6751]], device='cuda:0')
```

# 結論

src, tgtの値で同じ値と異なる値のところを比較したところ通常のtransformerの予想される動きと問題がない。よって、問題はtransformerのoutput内にあると考えられる。

**Comment by acesulfamK** [comment]
# 他の重みで確かめる
commit id: dbf2bd4a105b4d83802915fc7fa51327a340ce30

小さいデータセットで学習させた他のモデルにおいても、入力ごとに値が全く変わらないというエラーが発生した。

**Comment by acesulfamK** [comment]
# I maybe missed the architecture of Transformer



nn.Transformer implements only the part of whole transformer especially the repeating phase. So, it do not contain last Norm and Linear. There are 2 evidence.
- 1. In Transformer document, nn.Transformer have the output of the same length as tgt_input because of **"multi head attention"**. 
- 2. In someone's sample code, FNN is connected to the output of nn.Transformer.

Reference: 
- [nn.Transformer document](https://pytorch.org/docs/stable/generated/torch.nn.Transformer.html)
- [sample code](https://towardsdatascience.com/a-detailed-guide-to-pytorchs-nn-transformer-module-c80afbc9ffb1)

**Comment by acesulfamK** [comment]
Same output problem is solved. The problem that Training loss is too small compared to inference ability is handed over to #169
